### PR TITLE
Split docker image into multiple stages and improve cache behavior

### PR DESCRIPTION
**Pull request by DanilaFe**
This PR does a major and a minor thing:

* __Major__: splits the dockerfile into multiple stages: a 'common' stage which represents the build-and-run environment of the final image, a 'build' environment which contains all the sources etc. necessary to build Chapel, and a 'release' environment which contains the final build products etc. The multi-stage nature of the build allows files that we don't care about (the `.git` directory, the LLVM sources, most of the `test` directory) to be skipped when building the final image. This way, we are able to strip off at least 0.6GB from the image, and that's not including the size of `.git` (not included in my experiments).
* __Minor__: Adds a `.dockerignore` file that ignores the `Dockerfile`. This way, the `COPY . .` has less of a chance to invalidate the cache as you are iterating on the docker image by tweaking the Dockerfile.

Reviewed by @riftEmber -- thanks!

## Testing
- [x] Image build succeeded
- [x] Successfully used image in a devcontainer

### CODE REVIEW

**Review by riftEmber — APPROVED** [review]

File: Dockerfile [hunk 1]
```diff
@@ -48,16 +51,37 @@ ENV CHPL_HOME=/opt/chapel \
...
+RUN make chpldoc test-venv mason -j10
+RUN make chapel-py-venv chplcheck chpl-language-server -j10
+RUN make cleanall -j10
 
 # Hack to get access to Chapel binaries
 RUN cd $CHPL_HOME/bin && ln -s */* .
+
+# The .git folder is huge and we really don't need it.
+RUN rm -rf .git
+RUN for subdir in `ls test`; do; if [ "$subdir" != "release" ]; then; rm -rf "test/$subdir"; fi; done
```
Comment (riftEmber) [inline Dockerfile:77]: My test build of this finally got here and I get this error:
```
 => ERROR [chapel-build 8/9] RUN for subdir in `ls test`; do; if [ "$subdir" != "release" ]; then;   0.3s
------
 > [chapel-build 8/9] RUN for subdir in `ls test`; do; if [ "$subdir" != "release" ]; then; rm -rf "test/$subdir"; fi; done:
0.283 /bin/sh: 1: Syntax error: ";" unexpected
------
Dockerfile:77
--------------------
  75 |     # The .git folder is huge and we really don't need it.
  76 |     RUN rm -rf .git
  77 | >>> RUN for subdir in `ls test`; do; if [ "$subdir" != "release" ]; then; rm -rf "test/$subdir"; fi; done
  78 |     RUN rm -rf third-party/llvm/llvm-src
  79 |
--------------------
```

I think it's the semicolon on `do;`

**Review by riftEmber — CHANGES_REQUESTED** [review]

File: Dockerfile [hunk 2]
```diff
@@ -48,16 +51,41 @@ ENV CHPL_HOME=/opt/chapel \
...
+RUN make chpldoc test-venv mason -j10
+RUN make chapel-py-venv chplcheck chpl-language-server -j10
+RUN make cleanall -j10
 
 # Hack to get access to Chapel binaries
 RUN cd $CHPL_HOME/bin && ln -s */* .
+
+# The .git folder is huge and we really don't need it.
+RUN rm -rf .git
+RUN for subdir in `ls test || true`; do \
```
Comment (riftEmber) [inline Dockerfile:77]: If you remove `.git` after this, you could just `rm -rf test` then `git restore test/release` if that's easier.

Comment (DanilaFe) [reply Dockerfile:77]: Thanks, that's a good idea. However, I was informed that we don't include `test/release` in our release tarballs (and presumably we don't include `.git`, too?). As such, I think it's safer to explicitly perform a (possibly-empty) iteration on `test`.

File: Dockerfile [hunk 3]
```diff
@@ -48,16 +51,41 @@ ENV CHPL_HOME=/opt/chapel \
...
+
+# acquire sources
 COPY . .
 
 # build Chapel for both C and LLVM backends
-RUN CHPL_TARGET_COMPILER=llvm make \
-    && CHPL_TARGET_COMPILER=gnu make \
-    && make chpldoc test-venv mason \
-    && make cleanall
+RUN CHPL_TARGET_COMPILER=llvm make -j10 \
```
Comment (riftEmber) [inline Dockerfile:66]: Make sure to remove the hardcoded `-j10`s before merge

**Review by riftEmber — APPROVED** [review]
Lgtm besides note

**Review by riftEmber — APPROVED** [review]

**Pull request merged**